File: Java/Eta/Applications/Examples/src/main/java/com/refinitiv/eta/valueadd/examples/provider/DictionaryHandler.java
package com.refinitiv.eta.valueadd.examples.provider;

import com.refinitiv.eta.codec.Buffer;
import com.refinitiv.eta.codec.CodecFactory;
import com.refinitiv.eta.codec.CodecReturnCodes;
import com.refinitiv.eta.codec.DataDictionary;
import com.refinitiv.eta.codec.DataStates;
import com.refinitiv.eta.codec.EncodeIterator;
import com.refinitiv.eta.codec.StateCodes;
import com.refinitiv.eta.codec.StreamStates;
import com.refinitiv.eta.shared.DictionaryRejectReason;
import com.refinitiv.eta.shared.DictionaryRequestInfo;
import com.refinitiv.eta.shared.DictionaryRequestInfoList;
import com.refinitiv.eta.rdm.Dictionary;
import com.refinitiv.eta.transport.Error;
import com.refinitiv.eta.transport.TransportBuffer;
import com.refinitiv.eta.transport.TransportReturnCodes;
import com.refinitiv.eta.valueadd.domainrep.rdm.dictionary.DictionaryMsgFactory;
import com.refinitiv.eta.valueadd.domainrep.rdm.dictionary.DictionaryMsgType;
import com.refinitiv.eta.valueadd.domainrep.rdm.dictionary.DictionaryRefresh;
import com.refinitiv.eta.valueadd.domainrep.rdm.dictionary.DictionaryRefreshFlags;
import com.refinitiv.eta.valueadd.domainrep.rdm.dictionary.DictionaryRequest;
import com.refinitiv.eta.valueadd.domainrep.rdm.dictionary.DictionaryStatus;
import com.refinitiv.eta.valueadd.reactor.ReactorChannel;
import com.refinitiv.eta.valueadd.reactor.ReactorErrorInfo;
import com.refinitiv.eta.valueadd.reactor.ReactorFactory;
import com.refinitiv.eta.valueadd.reactor.ReactorSubmitOptions;

/*
 * This is the dictionary handler for the UPA - Java Interactive Provider
 * application. Only two dictionary streams (field and enum type) per channel
 * are allowed by this simple provider.
 * <p>
 * It provides methods for processing dictionary requests from consumers and
 * sending back the responses.
 * <p>
 * Methods for sending dictionary request reject/close status messages,
 * initializing the dictionary provider, loading dictionaries from files,
 * getting the loaded dictionary, and closing dictionary streams are also
 * provided.
 */
class DictionaryHandler
{
    private static final int MAX_FIELD_DICTIONARY_MSG_SIZE = 8192;
    private static final int MAX_ENUM_TYPE_DICTIONARY_MSG_SIZE = 12800;
    private static final int MAX_DICTIONARY_STATUS_MSG_SIZE = 1024;
    
    private static final String FIELD_DICTIONARY_FILE_NAME = "RDMFieldDictionary";
    private static final String ENUM_TABLE_FILE_NAME = "enumtype.def";

    // field dictionary name
    static final Buffer fieldDictionaryDownloadName;

    // enumtype dictionary name
    static final Buffer enumTypeDictionaryDownloadName;

    private ReactorSubmitOptions _submitOptions = ReactorFactory.createReactorSubmitOptions();

    static
    {
        fieldDictionaryDownloadName = CodecFactory.createBuffer();
        fieldDictionaryDownloadName.data("RWFFld");
        enumTypeDictionaryDownloadName = CodecFactory.createBuffer();
        enumTypeDictionaryDownloadName.data( "RWFEnum");
    }

    private DataDictionary _dictionary;
    private DictionaryRequestInfoList _dictionaryRequestInfoList;
    
    private EncodeIterator _encodeIter = CodecFactory.createEncodeIterator();
    private DictionaryRefresh _dictionaryRefresh = (DictionaryRefresh)DictionaryMsgFactory.createMsg();
    private DictionaryStatus _dictionaryStatus = (DictionaryStatus)DictionaryMsgFactory.createMsg();
    private DictionaryRequest _dictionaryRequest = (DictionaryRequest)DictionaryMsgFactory.createMsg();
    
    DictionaryHandler()
    {
        _dictionary = CodecFactory.createDataDictionary();
        _dictionaryRequestInfoList = new DictionaryRequestInfoList();
        _dictionaryRefresh.rdmMsgType(DictionaryMsgType.REFRESH);
        _dictionaryStatus.rdmMsgType(DictionaryMsgType.STATUS);
        _dictionaryRequest.rdmMsgType(DictionaryMsgType.REQUEST);
   }

    /*
     * Initializes dictionary information fields.
     */
    void init()
    {
        _dictionaryRequestInfoList.init();
    }

    /*
     * Returns the data dictionary.
     */
    DataDictionary dictionary()
    {
        return _dictionary;
    }

    /*
     * Calls the method to delete the dictionary, freeing all memory
     * associated with it.
     */
    void freeDictionary()
    {
        _dictionary.clear();
    }

    /*
     * Loads dictionary files
     */
    boolean loadDictionary(Error error)
    {
        _dictionary.clear();
        if (_dictionary.loadFieldDictionary(FIELD_DICTIONARY_FILE_NAME, error) < 0)
        {
            return false;
        }

        if (_dictionary.loadEnumTypeDictionary(ENUM_TABLE_FILE_NAME, error) < 0)
            return false;

        return true;
    }

    /*
     * Closes a dictionary stream.
     */
    void closeStream(int streamId)
    {
        // find original request information associated with streamId
        for (DictionaryRequestInfo dictionraryReqInfo : _dictionaryRequestInfoList)
        {
            if (dictionraryReqInfo.dictionaryRequest().streamId() == streamId && dictionraryReqInfo.isInUse())
            {
                // clear original request information
                System.out.println("Closing dictionary stream id " + dictionraryReqInfo.dictionaryRequest().streamId() +
                                   " with dictionary name: " + dictionraryReqInfo.dictionaryRequest().dictionaryName());
                dictionraryReqInfo.clear();
                break;
            }
        }
    }

    /*
     * Closes all open dictionary streams for a channel.
     */
    void closeStream(ReactorChannel chnl)
    {
        // find original request information associated with chnl
        for (DictionaryRequestInfo dictionraryReqInfo : _dictionaryRequestInfoList)
        {
            if (dictionraryReqInfo.channel() == chnl.channel() && dictionraryReqInfo.isInUse())
            {
                // clear original request information
                System.out.println("Closing dictionary stream id  " + dictionraryReqInfo.dictionaryRequest().streamId() +
                                   " with dictionary name: " + dictionraryReqInfo.dictionaryRequest().dictionaryName());
                dictionraryReqInfo.clear();
            }
        }
    }

    /*
     * 
     * Sends the dictionary close status message(s) for a channel. This consists
     * of finding all request information for this channel and sending the close
     * status messages to the channel.
     */
    int sendCloseStatusMsgs(ReactorChannel chnl, ReactorErrorInfo errorInfo)
    {
        int ret = 0;
        for (DictionaryRequestInfo dictionraryReqInfo : _dictionaryRequestInfoList)
        {
            if (dictionraryReqInfo.isInUse() && dictionraryReqInfo.channel() == chnl.channel())
            {
                ret = sendCloseStatusMsg(chnl, dictionraryReqInfo.dictionaryRequest().streamId(), errorInfo);
                if (ret != CodecReturnCodes.SUCCESS)
                    return ret;
            }
        }

        return CodecReturnCodes.SUCCESS;
    }

    /*
     * Sends the dictionary close status message for a channel. Returns success
     * if send dictionary close status succeeds or failure if it fails.
     */
    int sendCloseStatusMsg(ReactorChannel chnl, int streamId, ReactorErrorInfo errorInfo)
    {
        // get a buffer for the dictionary close status 
        TransportBuffer msgBuf = chnl.getBuffer(MAX_DICTIONARY_STATUS_MSG_SIZE, false, errorInfo);

        if (msgBuf != null)
        {
            // encode directory close
            _encodeIter.clear();
            int ret = _encodeIter.setBufferAndRWFVersion(msgBuf, chnl.majorVersion(), chnl.minorVersion());
            if (ret != CodecReturnCodes.SUCCESS)
            {
                errorInfo.error().text("EncodeIterator.setBufferAndRWFVersion() failed with return code: " + CodecReturnCodes.toString(ret));
                return ret;
            }
            
            // encode dictionary close status
            _dictionaryStatus.clear();
            _dictionaryStatus.streamId(streamId);
            _dictionaryStatus.applyHasState();
            _dictionaryStatus.state().streamState(StreamStates.CLOSED);
            _dictionaryStatus.state().dataState(DataStates.SUSPECT);
            _dictionaryStatus.state().text().data("Dictionary stream closed");
            ret = _dictionaryStatus.encode(_encodeIter);
            if (ret != CodecReturnCodes.SUCCESS)
            {
                System.out.println("encodeDictionaryCloseStatus() failed with return code: " + ret);
                return ret;
            }

            // send close status
            return chnl.submit(msgBuf, _submitOptions, errorInfo);
        }
        else
        {
            return CodecReturnCodes.FAILURE;
        }
    }

    /*
     * Sends the dictionary request reject status message for a channel. 
     */
    int sendRequestReject(ReactorChannel chnl, int streamId, DictionaryRejectReason reason, ReactorErrorInfo errorInfo)
    {
        // get a buffer for the dictionary request reject status
        TransportBuffer msgBuf = chnl.getBuffer(MAX_DICTIONARY_STATUS_MSG_SIZE, false, errorInfo);

        if (msgBuf != null)
        {
            // encode dictionary request reject status 
            int ret = encodeDictionaryRequestReject(chnl, streamId, reason, msgBuf, errorInfo);
            if (ret != CodecReturnCodes.SUCCESS)
            {
                return ret;
            }

            // send request reject status 
            return chnl.submit(msgBuf, _submitOptions, errorInfo);
        }
        else
        {
            return CodecReturnCodes.FAILURE;
        }
    }

    /*
     * Encodes the dictionary request reject status. Returns success if encoding
     * succeeds or failure if encoding fails. 
     */
    private int encodeDictionaryRequestReject(ReactorChannel chnl, int streamId, DictionaryRejectReason reason, TransportBuffer msgBuf, ReactorErrorInfo errorInfo)
    {
        // clear encode iterator
        _encodeIter.clear();

        // set-up message
        _dictionaryStatus.clear();
        _dictionaryStatus.streamId(streamId);
        _dictionaryStatus.applyHasState();
        _dictionaryStatus.state().dataState(DataStates.SUSPECT);

        switch (reason)
        {
            case UNKNOWN_DICTIONARY_NAME:
                _dictionaryStatus.state().code(StateCodes.NOT_FOUND);
                _dictionaryStatus.state().streamState(StreamStates.CLOSED);
                _dictionaryStatus.state().text().data("Dictionary request rejected for stream id " + streamId + " - dictionary name unknown");
                break;
            case MAX_DICTIONARY_REQUESTS_REACHED:
                _dictionaryStatus.state().code(StateCodes.TOO_MANY_ITEMS);
                _dictionaryStatus.state().streamState(StreamStates.CLOSED_RECOVER);
                _dictionaryStatus.state().text().data("Dictionary request rejected for stream id " + streamId + " -  max request count reached");
                break;
            case DICTIONARY_RDM_DECODER_FAILED:
                _dictionaryStatus.state().code(StateCodes.USAGE_ERROR);
                _dictionaryStatus.state().text().data("Dictionary request rejected for stream id " + streamId + " - decoding failure: " + errorInfo.error().text());
                break;
            default:
                break;
        }

        // encode message
        int ret = _encodeIter.setBufferAndRWFVersion(msgBuf, chnl.majorVersion(), chnl.minorVersion());
        if (ret != CodecReturnCodes.SUCCESS)
        {
            System.out.println("EncodeIterator.setBufferAndRWFVersion() failed with return code: " + CodecReturnCodes.toString(ret));
            return ret;
        }

        ret = _dictionaryStatus.encode(_encodeIter);
        if (ret != CodecReturnCodes.SUCCESS)
        {
            System.out.println("DictionaryStatus.encode() failed");
            return ret;
        }

        return CodecReturnCodes.SUCCESS;
    }


    /*
     * Sends a field dictionary response to a channel. This consists of getting
     * a message buffer, encoding the dictionary response, and sending the
     * dictionary response to the server. Returns success if send dictionary
     * response succeeds or failure if send response fails. 
     */
    int sendFieldDictionaryResponse(ReactorChannel chnl, DictionaryRequest dictionaryRequest, ReactorErrorInfo errorInfo)
    {
        _dictionaryRefresh.clear();

        _dictionaryRefresh.rdmMsgType(DictionaryMsgType.REFRESH);
        _dictionaryRefresh.streamId(dictionaryRequest.streamId());
        _dictionaryRefresh.dictionaryType(Dictionary.Types.FIELD_DEFINITIONS);
        _dictionaryRefresh.dictionary(_dictionary);
        _dictionaryRefresh.state().streamState(StreamStates.OPEN);
        _dictionaryRefresh.state().dataState(DataStates.OK);
        _dictionaryRefresh.state().code(StateCodes.NONE);
        _dictionaryRefresh.verbosity(dictionaryRequest.verbosity());
        _dictionaryRefresh.serviceId(dictionaryRequest.serviceId());
        _dictionaryRefresh.dictionaryName().data(dictionaryRequest.dictionaryName().data(),
                                                 dictionaryRequest.dictionaryName().position(), dictionaryRequest.dictionaryName().length());
        _dictionaryRefresh.applySolicited();
        
        boolean firstMultiPart = true;
        
        while (true)
        {
            // get a buffer for the dictionary response
            TransportBuffer msgBuf = chnl.getBuffer(MAX_FIELD_DICTIONARY_MSG_SIZE, false, errorInfo);
            if(msgBuf == null)
                return CodecReturnCodes.FAILURE;

            _dictionaryRefresh.state().text().data("Field Dictionary Refresh (starting fid " + _dictionaryRefresh.startFid() + ")");

            // clear encode iterator
            _encodeIter.clear();
            int ret = _encodeIter.setBufferAndRWFVersion(msgBuf, chnl.majorVersion(), chnl.minorVersion());
            if (ret != CodecReturnCodes.SUCCESS)
            {
                errorInfo.error().text("EncodeIterator.setBufferAndRWFVersion() failed with return code: " + CodecReturnCodes.toString(ret));
                return ret;
            }

            if (firstMultiPart)
            {
            	_dictionaryRefresh.applyClearCache();
            	firstMultiPart = false;
            }
            else
            	_dictionaryRefresh.flags( DictionaryRefreshFlags.SOLICITED );               
            
            // encode message
            ret = _dictionaryRefresh.encode(_encodeIter);
            if (ret < CodecReturnCodes.SUCCESS)
            {
                errorInfo.error().text("DictionaryRefresh.encode() failed");
                return ret;
            }

            // send dictionary response
            if (chnl.submit(msgBuf, _submitOptions, errorInfo) != TransportReturnCodes.SUCCESS)
                return CodecReturnCodes.FAILURE;

            // break out of loop when all dictionary responses sent
            if (ret == CodecReturnCodes.SUCCESS)
            {
                break;
            }

            // sleep between dictionary responses
            try
            {
                Thread.sleep(1);
            }
            catch (InterruptedException e)
            {
            }
        }
          
        return CodecReturnCodes.SUCCESS;
    }

    /*
     * Sends an enum type dictionary response to a channel. This consists of
     * getting a message buffer, encoding the dictionary response, and sending
     * the dictionary response to the server. Returns success if send dictionary
     * response succeeds or failure if send response fails. 
     */
    int sendEnumTypeDictionaryResponse(ReactorChannel chnl, DictionaryRequest dictionaryRequest, ReactorErrorInfo errorInfo)
    {
        _dictionaryRefresh.clear();
        
        _dictionaryRefresh.rdmMsgType(DictionaryMsgType.REFRESH);
        _dictionaryRefresh.streamId(dictionaryRequest.streamId());
        _dictionaryRefresh.dictionaryType(Dictionary.Types.ENUM_TABLES);
        _dictionaryRefresh.dictionary(_dictionary);
        _dictionaryRefresh.serviceId(dictionaryRequest.serviceId());
        _dictionaryRefresh.verbosity(dictionaryRequest.verbosity());
        _dictionaryRefresh.applySolicited();

        _dictionaryRefresh.state().streamState(StreamStates.OPEN);
        _dictionaryRefresh.state().dataState(DataStates.OK);
        _dictionaryRefresh.state().code(StateCodes.NONE);

        // dictionaryName
        _dictionaryRefresh.dictionaryName().data(dictionaryRequest.dictionaryName().data(),
                                                 dictionaryRequest.dictionaryName().position(), dictionaryRequest.dictionaryName().length());

        boolean firstMultiPart = true;
        
        while (true)
        {
            // get a buffer for the dictionary response
            TransportBuffer msgBuf = chnl.getBuffer(MAX_ENUM_TYPE_DICTIONARY_MSG_SIZE, false, errorInfo);
            if (msgBuf == null)
                return CodecReturnCodes.FAILURE;        	
        	
        	_dictionaryRefresh.state().text().data("Enum Type Dictionary Refresh (starting enum " + _dictionaryRefresh.startEnumTableCount() + ")");

        	// clear encode iterator
	        _encodeIter.clear();
	        int ret = _encodeIter.setBufferAndRWFVersion(msgBuf, chnl.majorVersion(), chnl.minorVersion());
	        if (ret < CodecReturnCodes.SUCCESS)
	        {
	            errorInfo.error().text("EncodeIterator.setBufferAndRWFVersion() failed with return code: " + CodecReturnCodes.toString(ret));
	            return ret;
	        }
	
            if (firstMultiPart)
            {
            	_dictionaryRefresh.applyClearCache();
            	firstMultiPart = false;
            }
            else
            	_dictionaryRefresh.flags( DictionaryRefreshFlags.SOLICITED );   	        
	        
	        // encode message
	        ret = _dictionaryRefresh.encode(_encodeIter);
	        if (ret < CodecReturnCodes.SUCCESS)
	        {
	            errorInfo.error().text("DictionaryRefresh.encode() failed");
	            return ret;
	        }
	
	        // send dictionary response
	        if (chnl.submit(msgBuf, _submitOptions, errorInfo) != TransportReturnCodes.SUCCESS)
	            return CodecReturnCodes.FAILURE;
	        
            // break out of loop when all dictionary responses sent
            if (ret == CodecReturnCodes.SUCCESS)
            {
                break;
            }

            // sleep between dictionary responses
            try
            {
                Thread.sleep(1);
            }
            catch (InterruptedException e)
            {
            }	        

        }

        return CodecReturnCodes.SUCCESS;
    }
    
    /*
     * Gets dictionary request information for the channel.
     */
    DictionaryRequestInfo getDictionaryRequestInfo(ReactorChannel reactorChannel, DictionaryRequest dictionaryRequest)
	{
		return _dictionaryRequestInfoList.get(reactorChannel.channel(), dictionaryRequest);
	}
}

File: CSharp/Ema/Src/Tests/OmmConsumerTests/DictionaryHandler.cs
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license      --
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
 *|                See the project's LICENSE.md for details.                  --
 *|           Copyright (C) 2023 Refinitiv. All rights reserved.              --
 *|-----------------------------------------------------------------------------
 */

using System;

using LSEG.Eta.Codec;
using LSEG.Eta.Common;
using LSEG.Eta.ValueAdd.Rdm;
using LSEG.Eta.ValueAdd.Reactor;

using static LSEG.Eta.Rdm.Dictionary;

namespace LSEG.Ema.Access.Tests.OmmConsumerTests;

/// Simple handler for <see cref="ProviderTest"/> to handle dictionary requests and
/// let <see cref="DictionaryCallbackClientTests"/> run its tests
internal class DictionaryHandler
{
    internal const string FIELD_DICTIONARY_NAME = "RWFFld";
    internal const string ENUM_TYPE_DICTIONARY_NAME = "RWFEnum";

    internal const string FIELD_DICTIONARY_FILENAME = "../../../ComplexTypeTests/RDMFieldDictionary";
    internal const string ENUM_TABLE_FILENAME = "../../../ComplexTypeTests/enumtype.def";

    private const int MAX_DICTIONARY_STATUS_MSG_SIZE = 1024;
    private const int MAX_FIELD_DICTIONARY_MSG_SIZE = 448000;
    private const int MAX_ENUM_TYPE_DICTIONARY_MSG_SIZE = 12800;

    private string FieldDictionaryDownloadName = FIELD_DICTIONARY_NAME;
    private string EnumTypeDictionaryDownloadName = ENUM_TYPE_DICTIONARY_NAME;

    private ReactorSubmitOptions m_SubmitOptions = new ReactorSubmitOptions();

    private EncodeIterator m_EncodeIter = new EncodeIterator();
    private DictionaryRefresh m_DictionaryRefresh = new DictionaryRefresh();
    private DictionaryStatus m_DictionaryStatus = new DictionaryStatus();

    internal DataDictionary DataDictionary { get; private set; } = new DataDictionary();
    private ProviderSessionOptions m_ProviderSessionOptions;

    public DictionaryHandler(ProviderTest providerTest)
    {
        m_ProviderSessionOptions = providerTest.ProviderSessionOptions;

        if (DataDictionary.LoadFieldDictionary(FIELD_DICTIONARY_FILENAME, out CodecError fdError) < 0)
        {
            Assert.Fail($"Failed to load FieldDictionary, error: {fdError?.Text}");
        }

        if (DataDictionary.LoadEnumTypeDictionary(ENUM_TABLE_FILENAME, out CodecError edError) < 0)
        {
            Assert.Fail($"Failed to load EnumType Dictionary, error: {edError?.Text}");
        }
    }

    public DictionaryHandler(ProviderSessionOptions providerSessionOptions)
    {
        m_ProviderSessionOptions = providerSessionOptions;

        if (DataDictionary.LoadFieldDictionary(FIELD_DICTIONARY_FILENAME, out CodecError fdError) < 0)
        {
            Assert.Fail($"Failed to load FieldDictionary, error: {fdError?.Text}");
        }

        if (DataDictionary.LoadEnumTypeDictionary(ENUM_TABLE_FILENAME, out CodecError edError) < 0)
        {
            Assert.Fail($"Failed to load EnumType Dictionary, error: {edError?.Text}");
        }
    }

    public ReactorReturnCode HandleDictionaryMsgEvent(RDMDictionaryMsgEvent reactorEvent)
    {
        DictionaryMsg dictionaryMsg = reactorEvent.DictionaryMsg!;
        ReactorChannel? reactorChannel = reactorEvent.ReactorChannel;

        ReactorReturnCode ret;

        switch (dictionaryMsg.DictionaryMsgType)
        {
            case DictionaryMsgType.REQUEST:
                {
                    DictionaryRequest dictionaryRequest = dictionaryMsg.DictionaryRequest!;

                    IRequestMsg requestMsg = (IRequestMsg)reactorEvent.Msg!;

                    if (!requestMsg.CheckNoRefresh()) // Check whether the consumer side requires a refresh
                    {
                        if (FieldDictionaryDownloadName.Equals(dictionaryRequest.DictionaryName.ToString()))
                        {
                            // Name matches field dictionary. Send the field dictionary refresh
                            if ((ret = SendFieldDictionaryResponse(reactorChannel!, dictionaryRequest)) != ReactorReturnCode.SUCCESS)
                            {
                                Assert.Fail($"SendFieldDictionaryResponse() failed: {ret}");
                            }
                        }
                        else if (EnumTypeDictionaryDownloadName.Equals(dictionaryRequest.DictionaryName.ToString()))
                        {
                            // Name matches the enum types dictionary. Send the enum types dictionary refresh
                            if ((ret = SendEnumTypeDictionaryResponse(reactorChannel!, dictionaryRequest)) != ReactorReturnCode.SUCCESS)
                            {
                                Assert.Fail($"SendEnumTypeDictionaryResponse() failed: {ret}");
                            }
                        }
                        else
                        {
                            if ((ret = SendRequestReject(reactorChannel!, reactorEvent.Msg!.StreamId)) != ReactorReturnCode.SUCCESS)
                            {
                                Assert.Fail($"SendRequestReject() failed: {ret}");
                            }
                        }
                    }
                    break;
                }

            case DictionaryMsgType.CLOSE:
                CloseStream(dictionaryMsg.StreamId);
                break;

            default:
                Assert.Fail($"Received Unhandled Dictionary Msg Type: {dictionaryMsg.DictionaryMsgType}");
                break;
        }

        return ReactorReturnCode.SUCCESS;
    }

    private void CloseStream(int streamId)
    {
        Assert.True(true);
    }

    private ReactorReturnCode SendRequestReject(ReactorChannel chnl, int streamId)
    {
        // get a buffer for the dictionary request reject status
        ITransportBuffer? msgBuf = chnl.GetBuffer(MAX_DICTIONARY_STATUS_MSG_SIZE, false, out _);

        if (msgBuf != null)
        {
            // encode dictionary request reject status
            ReactorReturnCode ret = EncodeDictionaryRequestReject(chnl, streamId, msgBuf);
            if (ret != ReactorReturnCode.SUCCESS)
            {
                return ret;
            }

            // send request reject status
            return chnl.Submit(msgBuf, m_SubmitOptions, out _);
        }
        else
        {
            return ReactorReturnCode.FAILURE;
        }
    }

    private ReactorReturnCode EncodeDictionaryRequestReject(ReactorChannel chnl, int streamId, ITransportBuffer msgBuf)
    {
        // clear encode iterator
        m_EncodeIter.Clear();

        // set-up message
        m_DictionaryStatus.Clear();
        m_DictionaryStatus.StreamId = streamId;
        m_DictionaryStatus.HasState = true;
        m_DictionaryStatus.State.DataState(DataStates.SUSPECT);

        // DictionaryRejectReason.UNKNOWN_DICTIONARY_NAME:
        m_DictionaryStatus.State.Code(StateCodes.NOT_FOUND);
        m_DictionaryStatus.State.StreamState(StreamStates.CLOSED);
        m_DictionaryStatus.State.Text().Data($"Dictionary request rejected for stream id {streamId} - dictionary name unknown");

        // encode message
        CodecReturnCode ret = m_EncodeIter.SetBufferAndRWFVersion(msgBuf, chnl.MajorVersion, chnl.MinorVersion);
        if (ret != CodecReturnCode.SUCCESS)
        {
            Assert.Fail($"EncodeIterator.SetBufferAndRWFVersion() failed with return code: {ret.GetAsString()}");
            return ReactorReturnCode.FAILURE;
        }

        ret = m_DictionaryStatus.Encode(m_EncodeIter);
        if (ret != CodecReturnCode.SUCCESS)
        {
            Assert.Fail("DictionaryStatus.Encode() failed");
            return ReactorReturnCode.FAILURE;
        }

        return ReactorReturnCode.SUCCESS;
    }

    private ReactorReturnCode SendEnumTypeDictionaryResponse(ReactorChannel chnl, DictionaryRequest dictionaryRequest)
    {
        m_DictionaryRefresh.Clear();

        m_DictionaryRefresh.StreamId = dictionaryRequest.StreamId;
        m_DictionaryRefresh.DictionaryType = Types.ENUM_TABLES;
        m_DictionaryRefresh.DataDictionary = DataDictionary;
        m_DictionaryRefresh.ServiceId = dictionaryRequest.ServiceId;
        m_DictionaryRefresh.Verbosity = dictionaryRequest.Verbosity;
        m_DictionaryRefresh.Solicited = true;

        m_DictionaryRefresh.State.StreamState(StreamStates.OPEN);
        m_DictionaryRefresh.State.DataState(DataStates.OK);
        m_DictionaryRefresh.State.Code(StateCodes.NONE);

        // dictionaryName
        m_DictionaryRefresh.DictionaryName = dictionaryRequest.DictionaryName;

        bool firstMultiPart = true;

        while (true)
        {
            // get a buffer for the dictionary response
            ITransportBuffer? msgBuf = chnl.GetBuffer(MAX_ENUM_TYPE_DICTIONARY_MSG_SIZE, false, out _);
            if (msgBuf == null)
            {
                return ReactorReturnCode.FAILURE;
            }

            m_DictionaryRefresh.State.Text().Data("Enum Type Dictionary Refresh (starting enum " + m_DictionaryRefresh.StartEnumTableCount + ")");

            msgBuf.Data.Limit = MAX_ENUM_TYPE_DICTIONARY_MSG_SIZE;

            // clear encode iterator
            m_EncodeIter.Clear();
            CodecReturnCode ret = m_EncodeIter.SetBufferAndRWFVersion(msgBuf, chnl.MajorVersion, chnl.MinorVersion);
            if (ret < CodecReturnCode.SUCCESS)
            {
                Assert.Fail($"EncodeIterator.SetBufferAndRWFVersion() failed with return code: {ret.GetAsString()}");
                return ReactorReturnCode.FAILURE;
            }

            if (firstMultiPart)
            {
                m_DictionaryRefresh.ClearCache = true;
                firstMultiPart = false;
            }
            else
                m_DictionaryRefresh.Flags = DictionaryRefreshFlags.SOLICITED;

            // encode message
            ret = m_DictionaryRefresh.Encode(m_EncodeIter);
            if (ret < CodecReturnCode.SUCCESS)
            {
                Assert.Fail("DictionaryRefresh.Encode() failed");
            }

            // send dictionary response
            ReactorReturnCode submitCode;
            if ((submitCode = chnl.Submit(msgBuf, m_SubmitOptions, out var submitError)) != ReactorReturnCode.SUCCESS)
            {
                Assert.Fail($"ReactorChannel.Submit failed ({submitCode}): {submitError}");
                return ReactorReturnCode.FAILURE;
            }

            // break out of loop when all dictionary responses sent
            if (ret == CodecReturnCode.SUCCESS)
            {
                break;
            }

            // sleep between dictionary responses
            try
            {
                System.Threading.Thread.Sleep(1);
            }
            catch (Exception)
            {
            }
        }

        return ReactorReturnCode.SUCCESS;
    }

    private ReactorReturnCode SendFieldDictionaryResponse(ReactorChannel chnl, DictionaryRequest dictionaryRequest)
    {
        m_DictionaryRefresh.Clear();

        m_DictionaryRefresh.StreamId = dictionaryRequest.StreamId;
        m_DictionaryRefresh.DictionaryType = Types.FIELD_DEFINITIONS;
        m_DictionaryRefresh.DataDictionary = DataDictionary;
        m_DictionaryRefresh.State.StreamState(StreamStates.OPEN);
        m_DictionaryRefresh.State.DataState(DataStates.OK);
        m_DictionaryRefresh.State.Code(StateCodes.NONE);
        m_DictionaryRefresh.Verbosity = dictionaryRequest.Verbosity;
        m_DictionaryRefresh.ServiceId = dictionaryRequest.ServiceId;
        m_DictionaryRefresh.DictionaryName = dictionaryRequest.DictionaryName;
        m_DictionaryRefresh.Solicited = true;

        bool firstMultiPart = true;

        while (true)
        {
            // get a buffer for the dictionary response
            ITransportBuffer? msgBuf = chnl.GetBuffer(MAX_FIELD_DICTIONARY_MSG_SIZE, false, out _);
            if (msgBuf == null)
            {
                Assert.True(false);
                return ReactorReturnCode.FAILURE;
            }

            m_DictionaryRefresh.State.Text().Data($"Field Dictionary Refresh (starting fid {m_DictionaryRefresh.StartFid})");

            msgBuf.Data.Limit = MAX_FIELD_DICTIONARY_MSG_SIZE;

            // clear encode iterator
            m_EncodeIter.Clear();
            CodecReturnCode ret = m_EncodeIter.SetBufferAndRWFVersion(msgBuf, chnl.MajorVersion, chnl.MinorVersion);
            if (ret != CodecReturnCode.SUCCESS)
            {
                Assert.Fail($"EncodeIterator.SetBufferAndRWFVersion() failed with return code: {ret.GetAsString()}");
                return ReactorReturnCode.FAILURE;
            }

            if (firstMultiPart)
            {
                m_DictionaryRefresh.ClearCache = true;
                firstMultiPart = false;
            }
            else
                m_DictionaryRefresh.Flags = DictionaryRefreshFlags.SOLICITED;

            // encode message
            ret = m_DictionaryRefresh.Encode(m_EncodeIter);
            if (ret < CodecReturnCode.SUCCESS)
            {
                Assert.Fail("DictionaryRefresh.Encode() failed");
                return ReactorReturnCode.FAILURE;
            }

            // send dictionary response
            if (chnl.Submit(msgBuf, m_SubmitOptions, out var submitError) != ReactorReturnCode.SUCCESS)
            {
                Assert.Fail($"ReactorChannel.Submit failed: {submitError}");
                return ReactorReturnCode.FAILURE;
            }

            // break out of loop when all dictionary responses sent
            if (ret == CodecReturnCode.SUCCESS)
            {
                break;
            }

            // sleep between dictionary responses
            try
            {
                System.Threading.Thread.Sleep(1);
            }
            catch (Exception)
            {
            }
        }

        return ReactorReturnCode.SUCCESS;
    }
}
